File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.HashMap;

class Solution {
    public int[] solution(int N, int[] stages) {
        //1. 스테이지별 도전자 수를 구한다.
        int[] challenger = new int[N + 2];
        for(int i = 0; i < stages.length; i++) {
            challenger[stages[i]] += 1;
        }
        
        //2. 스테이지별 실패한 사용자 수를 계산한다.
        HashMap<Integer, Double> fails = new HashMap<>();
        double total = stages.length;
        
        //3. 각 스테이지를 순회하며, 실패율 계산
        for(int i = 1; i <= N; i++) {
            if(challenger[i] == 0) { //4. 도전한 사람이 없는 경우
                fails.put(i, 0.); //실패율은 0
            }
            else {
                fails.put(i, challenger[i] / total); //5. 실패율을 구한다.
                total -= challenger[i]; //6. 다음 스테이지 실패율을 구하기 위해 현재 스테이지의 인원을 뺀다.
            }
        }
        //7. 실패율이 높은 스테이지부터 내림차순으로 정렬한다.
        return fails.entrySet().stream().sorted((o1, o2) -> Double.compare(o2.getValue(), o1.getValue())).mapToInt(HashMap.Entry::getKey).toArray();
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    // 각 스테이지별 사용자 수를 저장할 배열
    let user = new Array(N + 2).fill(0);
    
    for(let stage of stages) {
        user[stage]++;
    }
    
    // 실패율을 저장할 배열
    let failureRate = [];
    let total = stages.length;
    
    for(let i = 1; i <= N; i++) {
        if(total === 0) {
            failureRate.push({stage: i, rate: 0})
        } else {
            // 클리어하지 못한 사용자 수
            let unsuccessfulUser = user[i];
            failureRate.push({stage: i, rate: unsuccessfulUser / total})
            total -= unsuccessfulUser;
        }
    }
    failureRate.sort((a, b) => {
        if(b.rate !== a.rate) return b.rate - a.rate
        // 실패율이 같을 때 스테이지 번호가 낮은 순으로 정렬
        return a.stage - b.stage;
    })
    return failureRate.map(el => el.stage);
}